datastore/nats/nats.go

```diff
@@ -0,0 +1,161 @@
+package nats
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+	logrus "github.com/teslamotors/fleet-telemetry/logger"
+	"github.com/teslamotors/fleet-telemetry/metrics"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/server/airbrake"
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+// NatsConnect is a variable that holds the function to create a NATS connection
+// This allows for testing by replacing it with a mock
+var NatsConnect = nats.Connect
+
+// Producer client to handle NATS interactions
+type Producer struct {
+	natsConn           *nats.Conn
+	namespace          string
+	metricsCollector   metrics.MetricCollector
+	logger             *logrus.Logger
+	airbrakeHandler    *airbrake.Handler
+	ackChan            chan (*telemetry.Record)
+	reliableAckTxTypes map[string]interface{}
+}
+
+// Metrics stores metrics reported from this package
+type Metrics struct {
+	producerCount     adapter.Counter
+	bytesTotal        adapter.Counter
+	producerAckCount  adapter.Counter
+	bytesAckTotal     adapter.Counter
+	errorCount        adapter.Counter
+	reliableAckCount  adapter.Counter
+	producerQueueSize adapter.Gauge
+}
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// Config for NATS producer
+type Config struct {
+	URL string `json:"url"`
+}
+
+// NewProducer establishes the NATS connection and define the dispatch method
+func NewProducer(config *Config, namespace string, prometheusEnabled bool, metricsCollector metrics.MetricCollector, airbrakeHandler *airbrake.Handler, ackChan chan (*telemetry.Record), reliableAckTxTypes map[string]interface{}, logger *logrus.Logger) (telemetry.Producer, error) {
+	registerMetricsOnce(metricsCollector)
+
+	natsConn, err := NatsConnect(config.URL)
```

Comment:
I think it'd be a good idea to add some options:

```golang
nats.RetryOnFailedConnect(true), 
nats.MaxReconnects(-1), // so that it will not stop reconnecting
```
and potentially logging using
```
ClosedHandler
ErrorHandler
ConnectHandler
ReconnectErrHandler
DisconnectErrHandler
```

for visibility.